src/main/java/net/pms/dlna/DLNAResource.java

```diff
@@ -3958,53 +3726,438 @@ protected void setDiscovered(boolean discovered) {
+			}
+
+			if (configuration.isAutoloadExternalSubtitles()) {
+				SubtitleUtils.registerExternalSubtitles(file, media, forceRefresh);
+			}
+			List<DLNAMediaSubtitle> subtitlesList = media.getSubtitleTracksList();
+			if (subtitlesList != null) {
+				hasSubtitles = !subtitlesList.isEmpty();
+				for (DLNAMediaSubtitle subtitles : subtitlesList) {
+					if (subtitles.isExternal()) {
+						hasExternalSubtitles = true;
+						break;
+					}
+				}
+			}
+			isExternalSubtitlesParsed = true;
+		}
+	}
+
+	/**
+	 * Sets external subtitles parsed status to true and sets
+	 * {@link #hasSubtitles} and {@link #hasExternalSubtitles} according to the
+	 * existing {@link DLNAMediaSubtitle} instances.
+	 * <p>
+	 * <b>WARNING:</b> This should only be called when the subtitles tracks has
+	 * been populated by an alternative source like the database. Setting this
+	 * under other circumstances will break the implemented automatic parsing
+	 * and caching.
 	 */
-	protected void setHasExternalSubtitles(boolean value) {
-		this.hasExternalSubtitles = value;
+	public void setExternalSubtitlesParsed() {
+		if (media == null || !media.isVideo()) {
+			return;
+		}
+
+		synchronized (subtitlesLock) {
+			if (isExternalSubtitlesParsed) {
+				return;
+			}
+
+			List<DLNAMediaSubtitle> subtitlesList = media.getSubtitleTracksList();
+			hasSubtitles = !subtitlesList.isEmpty();
+			for (DLNAMediaSubtitle subtitles : subtitlesList) {
+				if (subtitles.isExternal()) {
+					hasExternalSubtitles = true;
+					break;
+				}
+			}
+			isExternalSubtitlesParsed = true;
+		}
+	}
+
+	/**
+	 * This method figures out which audio track should be used based on
+	 * {@link DLNAMediaInfo} metadata and configuration settings.
+	 *
+	 * @param renderer the {@link RendererConfiguration} from which to get the
+	 *            configuration or {@code null} to use the default
+	 *            configuration.
+	 * @return The resolved {@link DLNAMediaAudio} or {@code null}.
+	 */
+	public DLNAMediaAudio resolveAudioStream(RendererConfiguration renderer) {
+		if (media == null || media.getAudioTrackCount() == 0) {
+			LOGGER.trace("Found no audio track");
+			return null;
+		}
+		// Use device-specific pms conf
+		PmsConfiguration deviceSpecificConfiguration = PMS.getConfiguration(renderer);
+
+		// check for preferred audio
+		DLNAMediaAudio dtsTrack = null;
+		StringTokenizer st = new StringTokenizer(deviceSpecificConfiguration.getAudioLanguages(), ",");
+		while (st.hasMoreTokens()) {
+			String lang = st.nextToken().trim();
+			LOGGER.trace("Looking for an audio track with language \"{}\"", lang);
+			for (DLNAMediaAudio audio : media.getAudioTracksList()) {
+				if (audio.matchCode(lang)) {
+					LOGGER.trace("Matched audio track: {}", audio);
+					return audio;
+				}
+
+				if (dtsTrack == null && audio.isDTS()) {
+					dtsTrack = audio;
+				}
+			}
+		}
+
+		// preferred audio not found, take a default audio track, dts first if available
+		if (dtsTrack != null) {
```

Comment:
Also if it is not supported by the renderer ?

Reply:
@Sami32 I've just kept the current code, it might be a bug but I didn't really look at the audio tracks in this refactoring.